src/senselab/audio/tasks/features_extraction/praat_parselmouth.py

```diff
@@ -1,143 +1,1377 @@
+from senselab.utils.data_structures import logger
 
 
-def get_audios_durations(audios: List[Audio]) -> List[Dict[str, float]]:
-    """Returns the duration of the Audio objects."""
+def get_sound(audio: Union[Path, Audio], sampling_rate: int = 16000) -> parselmouth.Sound:
+    """Get a sound object from a given audio file or Audio object.
 
-    def get_audio_duration(audio: Audio) -> Dict[str, float]:
-        def _get_duration(sound: parselmouth.Sound) -> float:
-            return call(sound, "Get total duration")
+    Args:
+        audio (Union[Path, Audio]): A path to an audio file or an Audio object.
+        sampling_rate (int, optional): The sampling rate of the audio. Defaults to 16000.
 
-        waveform = audio.waveform
-        sampling_rate = audio.sampling_rate
-        sound = parselmouth.Sound(waveform, sampling_frequency=sampling_rate)
-        duration = _get_duration(sound)
-        return {"duration": duration}
+    Returns:
+        parselmouth.Sound: A Parselmouth Sound object.
 
-    durations = []
-    for audio in audios:
-        durations.append(get_audio_duration(audio))
-    return durations
+    Raises:
+        FileNotFoundError: If the file is not found at the given path.
+    """
+    try:
+        # Loading the sound
+        if isinstance(audio, Path):
+            audio = audio.resolve()
+            if not audio.exists():
+                logger.error(f"File does not exist: {audio}")
+                raise FileNotFoundError(f"File does not exist: {audio}")
+            snd_full = parselmouth.Sound(str(audio))
+        elif isinstance(audio, Audio):
+            snd_full = parselmouth.Sound(audio.waveform, audio.sampling_rate)
 
+        # Preprocessing
+        if parselmouth.praat.call(snd_full, "Get number of channels") > 1:
+            snd_full = snd_full.convert_to_mono()
+        if parselmouth.praat.call(snd_full, "Get sampling frequency") != sampling_rate:
+            snd_full = parselmouth.praat.call(snd_full, "Resample", sampling_rate, 50)
+            # Details of queery: https://www.fon.hum.uva.nl/praat/manual/Get_sampling_frequency.html
+            # Details of conversion: https://www.fon.hum.uva.nl/praat/manual/Sound__Resample___.html
+    except Exception as e:
+        raise RuntimeError(f"Error loading sound: {e}")
+    return snd_full
 
-def get_audios_f0_descriptors(
-    audios: List[Audio],
-    f0min: float,
-    f0max: float,
+
+def extract_speech_rate(snd: Union[parselmouth.Sound, Path, Audio]) -> Dict[str, float]:
+    """Extract speech timing and pausing features from a given sound object.
+
+    Args:
+        snd (Union[parselmouth.Sound, Path, Audio]): A Parselmouth Sound object or a file path or an Audio object.
+
+    Returns:
+        Dict[str, float]: A dictionary containing the following features:
+
+            - speaking_rate (float): Number of syllables divided by duration.
+            - articulation_rate (float): Number of syllables divided by phonation time.
```

Comment:
we should clarify what `syllable` means in the description as an acoustically defined syllable. this is an area of praat that means something quite specific based on acoustics and can miss several acoustic phonetic/linguistic considerations (e.g. long syllables, stress, etc.,.). it would be good to know if that has changed. also would this apply to whispered speech? i.e. would it return the same number of syllables between whispered and non-whispered speech of the same sentence?

Reply:
also pinging @nickcummins41 here.

Reply:
> we should clarify what syllable means in the description as an acoustically defined syllable. this is an area of praat that means something quite specific based on acoustics and can miss several acoustic phonetic/linguistic considerations (e.g. long syllables, stress, etc.,.). 

Good point, I can add that. The docs will definitely be refined once the paper we are writing is closer to completion. 


> also would this apply to whispered speech? i.e. would it return the same number of syllables between whispered and non-whispered speech of the same sentence?

Good question. It sounds like an interesting sub-project